CIPs/cip-0032.md

```diff
@@ -0,0 +1,117 @@
+
+We will introduce a new `AttestationSlasher` contract which allows slashing of stake if a completed attestation hasn't been observed in the past `B` blocks. To achieve this, we will upgrade the `Attestations` contract to register the selection and completion events in the new `AttestationSlasher` contract. The `AttestationSlasher` will contain the following state to track the latest block number that an attestation was requested/completed on:
+
+```js
+// Tracks the last time an attestation was requested from an issuer (AS node)
+mapping(address => uint256) public lastAttestationRequested;
+// Tracks the last time an attestation was completed by an issuer
+mapping(address => uint256) public lastAttestationCompleted;
+// Tracks the last time an issuer was last slashed
+mapping(address => uint256) public lastSlashed;
+```
+
+A slashing event may occur when the slashing initiator calls `slash()` with the slashable address. Before slashing, the slashing contract checks the following:
+
+1. The validator has been elected continuously for the past `B` blocks.
+2. The issuer was selected for at least one attestation in the past `B` blocks.
+3. The validator wasn't slashed by the `AttestationSlasher` within the past `B` blocks.
+4. The address hasn't completed an attestation in the past `B` blocks.
+5. The addres hasn't submitted `SignalOnline` event in the last `S` blocks. (see [below](#signal-online)).
+
+If the validator was intermittently elected, the contract will only consider the past `B` blocks in which the validator was *continuously* elected. For example, if a validator with a downed attestation node was elected for .5 `B` blocks, was unelected, and re-elected for .5 `B` blocks, then it will not be eligible for slashing since the slasher restarts counting when the validator was re-elected. Additionally, if the validator was elected but was never selected to provide an attestation in the entire period of `B` blocks, then they are not eligible for slashing.
+
+### Signal online
+
+In the unlikely event that a validator was never selected for an attestation during the period of `B` blocks, it's possible for a malicious actor to request multiple attestations with the goal of randomly getting paired with this validator and intentionally leaving the attestation incomplete. To prevent this, we're building into the AS the ability to signal to the `AttestationSlasher` that the service is still online if a slashing event becomes close by signing and submitting a transaction to the slasher contract. This requires that the AS signer key to be minimally funded to pay the gas fees of the transaction, however this adds more piece of mind for validators to be sure they won't be incorrectly slashed. Additionally, we'll be adding two new Prometheus events to the AS to alert a validator when the slashing period is nearly reached and another event to alert when the AS signer key's balance is low.
```

Comment:
Seems like one of the main reasons for signalOnline is to handle the case of non completed attestations from a "down" user.  (e.g. the cause of a non completed attestation is the user's fault, as opposed to the AS's).

I am not very familiar with the attestation protocol, so this may not be technically feasible, but perhaps one heuristic to figure out if it's more likely that it's the user's fault is if they didn't complete any other attestations (I believe 3 completed is required by the user?).

Ofc this is just a heuristic (e.g. the user could be malicious and intentionally not complete an attestation for a specific validator), so if it is more likely the user, then perhaps the validator is only fractionally slashed (or something along those lines).

Reply:
In practice, it sounds more like "if the AS appears to be available it doesn't get penalized". In this case maybe we could track the ratio of the AS being selected to the number of "initialized attestations" instead?

Reply:
@kevjue I was also thinking along the same lines and already added it [our dashboard](https://metabase.celo-networks-dev.org/public/dashboard/b0a27650-1d62-4645-81d7-26ff7546ff0d) (see `OnlyMissingIssuer` column in the `Missed Attestations` table). `OnlyMissingIssuer` implies that this validator was the only one that the user missed when completing attestations, indicating it's likely the fault of the validator and not the user. This is metric is nice because it is derived purely from on-chain data. Once we're more confident in the metric we can use this to reward validators who run with exceptional services.

The SignalOnline feature does help cover this gap, but it also can cover cases of an outage caused by cLabs (ex. Valora or Komenci are having issues), or simply unexpected lulls in user onboarding. Once we build more confidence in this min-bar metric, we can remove the SignalOnline feature.

Reply:
@carterqw2 by "initialized attestations" do you mean the number of attestations that were completed for that issuer? The concern with ratios for now is that they're gamable by malicious actors (ex. I can make other validators look bad or make my own look good). In addition to the metric that Kevin called out, I'd like to also explore adding an on-chain attestation from Komenci indicating that the user solved a reCaptcha and device checks. This attestation can be used to build confidence in the on-chain metric and make it more difficult to game.